src/Data/Graph.purs

```diff
@@ -1,152 +1,107 @@
 -- | A data structure and functions for graphs
 
 module Data.Graph
-  ( Edge(..)
-  , Graph(..)
-  , SCC(..)
+  ( Graph
+  , unfoldGraph
   , vertices
-  , scc
-  , scc'
-  , topSort
-  , topSort'
+  , lookup
+  , outEdges
+  , topologicalSort
   ) where
 
-import Prelude hiding
-
-import Control.Monad.Eff (runPure)
-import Control.Monad.ST (writeSTRef, modifySTRef, readSTRef, newSTRef, runST)
-
-import Data.Foldable (any, for_, elem)
-import Data.List (List(..), concatMap, reverse, singleton)
+import Prelude
+import Data.CatList (CatList)
+import Data.CatList as CL
+import Data.Foldable (class Foldable)
+import Data.List (List(..))
+import Data.List as L
+import Data.Map (Map)
 import Data.Map as M
-import Data.Maybe (Maybe(..), isNothing)
-import Data.Traversable (for)
-
--- | An directed edge between vertices labelled with keys of type `k`.
-data Edge k = Edge k k
+import Data.Maybe (Maybe(..), maybe)
+import Data.Monoid (mempty)
+import Data.Tuple (Tuple(..))
 
 -- | A graph with vertices of type `v`.
 -- |
 -- | Edges refer to vertices using keys of type `k`.
-data Graph k v = Graph (List v) (List (Edge k))
-
--- | A strongly-connected component of a graph.
+newtype Graph k v = Graph
+  { vertices :: Map k { value :: v, out :: List k }
```

Comment:
Can we not just newtype the `Map` directly? It'd be more efficient than having the intermediate record, and since there's only one field it seems a little redundant.